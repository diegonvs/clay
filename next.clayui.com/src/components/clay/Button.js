/**
 * Â© 2019 Liferay, Inc. <https://liferay.com>
 *
 * SPDX-License-Identifier: BSD-3-Clause
 */

import ClayButton from '@clayui/button';
import ClayIcon from '@clayui/icon';
import Editor from './Editor';
import React from 'react';

const spritemap = '/images/icons/icons.svg';

export const ButtonBlock = () => {
	const scope = {ClayButton, spritemap};
	const code = `const Component = () => {

	return (
        <ClayButton displayType="primary" block>
			Button Block
		</ClayButton>
	);
}

render(<Component />)`;

	return <Editor code={code} scope={scope} />;
};

export const ButtonDisplayTypes = () => {
	const scope = {ClayButton, spritemap};
	const code = `const Component = () => {

	return (
		<>
			<ClayButton displayType="primary">
				Button Primary
			</ClayButton>
			<ClayButton displayType="secondary">
				Button Secondary
			</ClayButton>
			<ClayButton displayType="link">
				Button Link
			</ClayButton>
			<ClayButton displayType="unstyled">
				Button Unstyled
			</ClayButton>
		</>
	);
}

render(<Component />)`;

	return <Editor code={code} scope={scope} />;
};

export const ButtonGroup = () => {
	const scope = {ClayButton, spritemap};
	const code = `const Component = () => {

	return (
        <ClayButton.Group>
			<ClayButton>{'This'}</ClayButton>
			<ClayButton displayType="secondary">{'is'}</ClayButton>
			<ClayButton>{'a'}</ClayButton>
			<ClayButton displayType="secondary">{'button'}</ClayButton>
			<ClayButton>{'group.'}</ClayButton>
		</ClayButton.Group>
	);
}

render(<Component />)`;

	return <Editor code={code} scope={scope} />;
};

export const ButtonMonospaced = () => {
	const scope = {ClayButton, ClayIcon, spritemap};
	const code = `const Component = () => {

	return (
		<ClayButton displayType="primary" monospaced>
			<ClayIcon spritemap={spritemap} symbol="plus"></ClayIcon>
		</ClayButton>
	);
}

render(<Component />)`;

	return <Editor code={code} scope={scope} />;
};

export const ButtonWithStar = () => {
	const scope = {ClayButton, ClayIcon, spritemap};
	const code = `const Component = () => {

	return (
		<ClayButton displayType="primary">
			<ClayIcon spritemap={spritemap} symbol="bookmarks"></ClayIcon> Bookmark
		</ClayButton>
	);
}

render(<Component />)`;

	return <Editor code={code} scope={scope} />;
};

export const ButtonSmall = () => {
	const scope = {ClayButton, ClayIcon, spritemap};
	const code = `const Component = () => {

	return (
		<ClayButton displayType="primary" small>
			<ClayIcon spritemap={spritemap} symbol="plus"></ClayIcon>
		</ClayButton>
	);
}

render(<Component />)`;

	return <Editor code={code} scope={scope} />;
};
