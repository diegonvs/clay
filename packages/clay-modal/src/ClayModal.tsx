/**
 * Â© 2019 Liferay, Inc. <https://liferay.com>
 *
 * SPDX-License-Identifier: BSD-3-Clause
 */

import Button from '@clayui/button';
import classNames from 'classnames';
import Icon from '@clayui/icon';
import React, {FunctionComponent, useEffect, useRef, useState} from 'react';
import {createPortal} from 'react-dom';
import {Footer, FooterContainer, FooterContainerProps} from './Footer';
import {useModal} from './Hook';

type Size = 'full-screen' | 'lg' | 'sm';

type Status = 'danger' | 'info' | 'success' | 'warning';

interface Props extends React.HTMLAttributes<HTMLDivElement> {
	/**
	 * Callback called to close the modal.
	 */
	onClose: () => void;

	/**
	 * Renders the footer with the FooterContainer component utilities.
	 */
	renderFooter?: React.ReactElement;

	/**
	 * The size of element modal.
	 */
	size?: Size;

	/**
	 * The path to the SVG spritemap file containing the icons.
	 */
	spritemap?: string;

	/**
	 * Status messages.
	 */
	status?: Status;

	/**
	 * Title of the modal.
	 */
	title?: string;

	/**
	 * Url to place an iframe in the body of the modal.
	 */
	url?: string;

	/**
	 * Flag to indicate if the modal is open or closed.
	 */
	visible: boolean;
}

const ICON_MAP = {
	danger: 'exclamation-full',
	info: 'info-circle',
	success: 'check-circle',
	warning: 'question-circle-full',
};

const Portal = ({children}: {children: React.ReactNode}) => {
	const portalRef = useRef(document.createElement('div'));
	const elToMountTo = document && document.body;

	useEffect(() => {
		elToMountTo.appendChild(portalRef.current);
		return () => {
			elToMountTo.removeChild(portalRef.current);
		};
	}, [elToMountTo]);

	return createPortal(children, portalRef.current);
};

const delay = (fn: Function) => {
	const callbackId = setTimeout(() => {
		fn();
		clearTimeout(callbackId);
	}, 100);
};

const ClayModal: FunctionComponent<Props> & {
	FooterContainer: FunctionComponent<FooterContainerProps>;
} = ({
	children,
	className,
	onClose = () => {},
	renderFooter,
	size,
	spritemap,
	status,
	title,
	url,
	visible = false,
	...otherProps
}) => {
	/**
	 * It is set to keep track of modal visibility and ensure that classes
	 * are removed when necessary.
	 */
	const [visibleModal, setVisibleModal] = useState<boolean>(visible);

	const [visibleClassShow, setVisibleClassShow] = useState<boolean>(false);

	const modalDialogElementRef = useRef<HTMLDivElement | null>(null);

	/**
	 * A hook that takes care of controlling click, keyup and keydown events
	 * respectively close the modal after a click on the overlay, close the
	 * modal by pressing the ESC key and control the focus within the Modal.
	 */
	useModal(modalDialogElementRef, visible, onClose);

	/**
	 * Synchronize the visible of the props with the internal value to update
	 * the states with delay, this ensures that the animation occurs when
	 * opening and closing the modal.
	 */
	useEffect(() => {
		if (visible) {
			document.body.classList.add('modal-open');
			setVisibleModal(visible);
			delay(() => setVisibleClassShow(visible));
		} else {
			document.body.classList.remove('modal-open');
			setVisibleClassShow(visible);
			delay(() => setVisibleModal(visible));
		}
	}, [visible]);

	return (
		<>
			{visibleModal && (
				<Portal>
					<div
						className={classNames('modal-backdrop fade', {
							show: visibleClassShow,
						})}
					/>
				</Portal>
			)}
			<div
				{...otherProps}
				className={classNames('fade modal', className, {
					'd-block': visibleModal,
					show: visibleClassShow,
				})}
			>
				<div
					className={classNames('modal-dialog', {
						[`modal-${size}`]: size,
						[`modal-${status}`]: status,
					})}
					ref={modalDialogElementRef}
					tabIndex={-1}
				>
					<div className="modal-content">
						{title && (
							<div className="modal-header">
								{status && (
									<div className="modal-title-indicator">
										<Icon
											spritemap={spritemap}
											symbol={ICON_MAP[status]}
										/>
									</div>
								)}
								<div className="modal-title">{title}</div>
								<Button
									aria-label="close"
									className="close"
									displayType="unstyled"
									onClick={onClose}
								>
									<Icon
										spritemap={spritemap}
										symbol="times"
									/>
								</Button>
							</div>
						)}
						<div
							className={classNames('modal-body', {
								'modal-body-iframe': url,
							})}
						>
							{url ? <iframe src={url} title={url} /> : children}
						</div>
						{renderFooter && <Footer>{renderFooter}</Footer>}
					</div>
				</div>
			</div>
		</>
	);
};

ClayModal.FooterContainer = FooterContainer;

export default ClayModal;
