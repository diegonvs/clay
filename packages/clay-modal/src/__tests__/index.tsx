/**
 * Â© 2019 Liferay, Inc. <https://liferay.com>
 *
 * SPDX-License-Identifier: BSD-3-Clause
 */

import * as React from 'react';
import * as TestRenderer from 'react-test-renderer';
import Button from '@clayui/button';
import ClayModal from '..';
import ReactDOM from 'react-dom';

const spritemap = 'icons.svg';

describe('ClayModal', () => {
	beforeAll(() => {
		// @ts-ignore
		ReactDOM.createPortal = jest.fn((element, node) => {
			return element;
		});
	});

	afterEach(() => {
		jest.restoreAllMocks();
	});

	it('renders', () => {
		const testRenderer = TestRenderer.create(
			<ClayModal onClose={() => {}} spritemap={spritemap}>
				{onClose => (
					<>
						<ClayModal.Header>{'Foo'}</ClayModal.Header>
						<ClayModal.Body>
							<h1>{'Hello world!'}</h1>
						</ClayModal.Body>
						<ClayModal.Footer
							last={
								<Button onClick={onClose}>{'Primary'}</Button>
							}
						/>
					</>
				)}
			</ClayModal>
		);

		expect(testRenderer.toJSON()).toMatchSnapshot();
	});

	it('renders with Header', () => {
		const testRenderer = TestRenderer.create(
			<ClayModal onClose={() => {}} spritemap={spritemap}>
				{() => <ClayModal.Header>{'Foo'}</ClayModal.Header>}
			</ClayModal>
		);

		expect(testRenderer.toJSON()).toMatchSnapshot();
	});

	it('renders with size', () => {
		const testRenderer = TestRenderer.create(
			<ClayModal onClose={() => {}} size="lg" spritemap={spritemap} />
		);

		expect(testRenderer.toJSON()).toMatchSnapshot();
	});

	it('renders with status', () => {
		const testRenderer = TestRenderer.create(
			<ClayModal
				onClose={() => {}}
				spritemap={spritemap}
				status="success"
			/>
		);

		expect(testRenderer.toJSON()).toMatchSnapshot();
	});

	it('renders a body component with url', () => {
		const testRenderer = TestRenderer.create(
			<ClayModal onClose={() => {}} spritemap={spritemap}>
				{() => <ClayModal.Body url="http://localhost" />}
			</ClayModal>
		);

		expect(testRenderer.toJSON()).toMatchSnapshot();
	});

	it('renders a footer component with buttons', () => {
		const testRenderer = TestRenderer.create(
			<ClayModal onClose={() => {}} spritemap={spritemap}>
				{onClose => (
					<ClayModal.Footer
						first={<Button>{'Bar'}</Button>}
						last={<Button onClick={onClose}>{'Foo'}</Button>}
						middle={<Button>{'Baz'}</Button>}
					/>
				)}
			</ClayModal>
		);

		expect(testRenderer.toJSON()).toMatchSnapshot();
	});
});
