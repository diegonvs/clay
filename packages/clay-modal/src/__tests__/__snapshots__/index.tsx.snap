// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClayModal renders 1`] = `
Array [
  <div
    className="modal-backdrop fade"
  />,
  <div
    className="fade modal d-block"
  >
    <div
      className="modal-dialog"
      tabIndex={-1}
    >
      <div
        className="modal-content"
      >
        <div
          className="modal-header"
        >
          <div
            className="modal-title"
          >
            Foo
          </div>
          <button
            aria-label="close"
            className="close btn btn-unstyled"
            onClick={[Function]}
            type="button"
          >
            <svg
              className="lexicon-icon lexicon-icon-times"
              role="presentation"
            >
              <use
                xlinkHref="icons.svg#times"
              />
            </svg>
          </button>
        </div>
        <div
          className="modal-body"
        >
          <h1>
            Hello world!
          </h1>
        </div>
        <div
          className="modal-footer"
        >
          <div
            className="modal-item-first"
          />
          <div
            className="modal-item"
          />
          <div
            className="modal-item-last"
          >
            <button
              className="btn btn-primary"
              onClick={[Function]}
              type="button"
            >
              Primary
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[`ClayModal renders a body component with url 1`] = `
Array [
  <div
    className="modal-backdrop fade"
  />,
  <div
    className="fade modal d-block"
  >
    <div
      className="modal-dialog"
      tabIndex={-1}
    >
      <div
        className="modal-content"
      >
        <div
          className="modal-body modal-body-iframe"
        >
          <iframe
            src="http://localhost"
            title="http://localhost"
          />
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[`ClayModal renders a footer component with buttons 1`] = `
Array [
  <div
    className="modal-backdrop fade"
  />,
  <div
    className="fade modal d-block"
  >
    <div
      className="modal-dialog"
      tabIndex={-1}
    >
      <div
        className="modal-content"
      >
        <div
          className="modal-footer"
        >
          <div
            className="modal-item-first"
          >
            <button
              className="btn btn-primary"
              type="button"
            >
              Bar
            </button>
          </div>
          <div
            className="modal-item"
          >
            <button
              className="btn btn-primary"
              type="button"
            >
              Baz
            </button>
          </div>
          <div
            className="modal-item-last"
          >
            <button
              className="btn btn-primary"
              onClick={[Function]}
              type="button"
            >
              Foo
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[`ClayModal renders with Header 1`] = `
Array [
  <div
    className="modal-backdrop fade"
  />,
  <div
    className="fade modal d-block"
  >
    <div
      className="modal-dialog"
      tabIndex={-1}
    >
      <div
        className="modal-content"
      >
        <div
          className="modal-header"
        >
          <div
            className="modal-title"
          >
            Foo
          </div>
          <button
            aria-label="close"
            className="close btn btn-unstyled"
            onClick={[Function]}
            type="button"
          >
            <svg
              className="lexicon-icon lexicon-icon-times"
              role="presentation"
            >
              <use
                xlinkHref="icons.svg#times"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[`ClayModal renders with size 1`] = `
Array [
  <div
    className="modal-backdrop fade"
  />,
  <div
    className="fade modal d-block"
  >
    <div
      className="modal-dialog modal-lg"
      tabIndex={-1}
    >
      <div
        className="modal-content"
      />
    </div>
  </div>,
]
`;

exports[`ClayModal renders with status 1`] = `
Array [
  <div
    className="modal-backdrop fade"
  />,
  <div
    className="fade modal d-block"
  >
    <div
      className="modal-dialog modal-success"
      tabIndex={-1}
    >
      <div
        className="modal-content"
      />
    </div>
  </div>,
]
`;
